File: backend/api/app/domain/use_cases/user/create.py
from api.core.use_cases.use_case import UseCase
from api.app.domain.repositories.user import BaseRepository
from api.core.common.equatable import Equatable
from core.common.either import Either
from core.errors.failure import Failure
from app.domain.entities.user import User, UserEntity

class Params(Equatable):
    def __init__(self, user: UserEntity) -> None:
        self.user = user


class CreateUser(UseCase[User]):
    def __init__(self, repository: BaseRepository):
        self.repository = repository
    
    async def __call__(self, params: Params) -> Either[Failure, User]:
        return await self.repository.create_user(params.user)

File: mobile/lib/features/architect/domains/use_cases/post/create.dart
import 'package:dartz/dartz.dart';
import 'package:equatable/equatable.dart';

import '../../../../../core/errors/failure.dart';
import '../../../../../core/use_cases/usecase.dart';
import '../../entities/post.dart';
import '../../repositories/post.dart';

class CreatePost implements UseCase<Post, Params> {
  final PostRepository repository;

  const CreatePost(this.repository);

  @override
  Future<Either<Failure, Post>> call(Params params) async {
    return await repository.create(
      image: params.image,
      title: params.title,
      content: params.content,
      userId: params.userId,
      tags: params.tags,
    );
  }
}

class Params extends Equatable {
  final String image;
  final String title;
  final String? content;
  final String userId;
  final List<String> tags;

  const Params({
    required this.image,
    required this.title,
    this.content,
    required this.userId,
    required this.tags,
  });

  @override
  List<Object?> get props => [image, title, content, userId, tags];
}
